drivers/rilmodem/sim.c

```diff
@@ -640,6 +640,60 @@ static void configure_active_app(struct sim_data *sd,
 	}
 }
 
+static void sim_send_set_uicc_subscription(struct ofono_sim *sim, int slot_id, int app_index, int sub_id, int sub_status)
+{
+    struct sim_data *sd = ofono_sim_get_data(sim);
+    struct parcel rilp;
+
+    DBG("");
+
+    g_ril_request_set_uicc_subscription(sd->ril, slot_id, app_index, sub_id, sub_status, &rilp);
+    g_ril_send(sd->ril, RIL_REQUEST_SET_UICC_SUBSCRIPTION, &rilp,
+               NULL, NULL, NULL);
+}
+
+static int sim_select_uicc_subscription(struct ofono_sim *sim, struct reply_sim_status *status)
+{
+    struct sim_data *sd = ofono_sim_get_data(sim);
+    int slot_id = ofono_modem_get_integer(sd->modem, "Slot");
+    int selected_app = -1;
+    unsigned int i;
+
+    for (i=0; i < status->num_apps; i++) {
+        switch (status->apps[i]->app_type) {
+        case RIL_APPTYPE_UNKNOWN:
+            continue;
+        case RIL_APPTYPE_USIM:
+        case RIL_APPTYPE_RUIM:
+            if (selected_app != -1) {
+                switch (status->apps[selected_app]->app_type) {
+                case RIL_APPTYPE_USIM:
+                case RIL_APPTYPE_RUIM:
+                    break;
+                default:
+                    selected_app = i;
```

Comment:
This is algorithm used in CM [1]. It prefers USIM | RUIM over SIM | CSIM. (That second app_type check replaces selected_app with current one if current selected_app isn't USIM | RUIM) As there isn't CDMA support in oFono yet, it may be make sense to just log that and skip that app.
[1] https://github.com/CyanogenMod/android_frameworks_opt_telephony-msim/blob/cm-11.0/frameworks/src/com/codeaurora/telephony/msim/SubscriptionManager.java#L401-L450